tools/harness-simulation/posix/launch_testbed.py

```diff
@@ -131,53 +128,41 @@ def main():
 
     # Parse arguments
     parser = argparse.ArgumentParser()
-
-    # Determine the interface
-    parser.add_argument('-i',
-                        '--interface',
-                        dest='ifname',
+    parser.add_argument('-c',
+                        '--config',
+                        dest='config',
                         type=str,
                         required=True,
-                        help='the interface used for discovery')
-
-    # Determine the number of OpenThread FTD simulations to be "detected" and then started
-    parser.add_argument('--ot',
-                        dest='ot_num',
-                        type=int,
-                        required=False,
-                        default=0,
-                        help=f'the number of OpenThread FTD simulations')
-
-    # Determine the number of OpenThread BR simulations to be initiated and then detected
-    parser.add_argument('--otbr',
-                        dest='otbr_num',
-                        type=int,
-                        required=False,
-                        default=0,
-                        help=f'the number of OpenThread BR simulations')
-
-    # Determine the number of sniffer simulations to be started and then detected
-    parser.add_argument('--sniffer',
-                        dest='sniffer_num',
-                        type=int,
-                        required=False,
-                        default=0,
-                        help=f'the number of sniffer simulations, no more than {MAX_SNIFFER_NUM}')
-
+                        help='the path of the configuration JSON file')
     args = parser.parse_args()
+    with open(args.config, 'r') as f:
+        config = json.load(f)
+
+    ot_path = config['ot_path']
+    ot_build = config['ot_build']
+    max_nodes_num = ot_build['max_number']
+    # No test case requires more than 2 sniffers
+    max_sniffer_num = 2
+
+    ot_devices = [(item['tag'], item['number']) for item in ot_build['ot']]
+    otbr_devices = [(item['tag'], item['number']) for item in ot_build['otbr']]
+    ot_nodes_num = sum(x[1] for x in ot_devices)
+    otbr_nodes_num = sum(x[1] for x in otbr_devices)
+    nodes_num = ot_nodes_num + otbr_nodes_num
+    sniffer_num = config['sniffer']['number']
 
-    # Check validation of arguments
-    if args.ot_num < 0:
-        raise ValueError(f'The number of FTDs should be non-negative')
+    # Check validation of numbers
+    if not all(0 <= x[1] <= max_nodes_num for x in ot_devices):
+        raise ValueError(f'The number of devices of each OT version should be between 0 and {max_nodes_num}')
 
-    if args.otbr_num < 0:
-        raise ValueError(f'The number of OTBRs should be non-negative')
+    if not all(0 <= x[1] <= max_nodes_num for x in otbr_devices):
+        raise ValueError(f'The number of devices of each OTBR version should be between 0 and {max_nodes_num}')
```

Comment:
I think there should be different limits for OTBR and OT devices. What factors are considered when defining the limits?

Reply:
The check here only limits the number should be non-negative. The user can start whatever the number of devices regardless of their types. The total number of devices should not exceed `max_nodes_num`, because `max_nodes_num` also controls the build argument `OT_SIMULATION_MAX_NETWORK_SIZE`, which defines the maximum allowed node ID.

Reply:
Make sense to check for negative numbers.